### DISCUSSION TITLE
How to add a delay to Person Detected

### DISCUSSION THREAD

**Discussion by Zipties** [TheRealWaldo/AMG8833-ESPHOME#9] [Q&A]
Every 6 seconds my sensor flaps between detecting a person, to not detecting a person for under a second, and then back to the person detected for 6 more seconds. 

What is the best way to handle that? I was thinking a simple 1-second delay before switching states would work, but I am not really sure where to define that.

```
esphome:
  name: upstairs-grid-eye
  includes:
  - amg8833.h
  - amg8833_camera.h
  - amg8833_person_detected.h
  libraries:
    - "Wire"
    - "SparkFun GridEYE AMG88 Library"

esp32:
  board: esp32dev
  framework:
    type: arduino

# Enable logging
logger:

# Enable Home Assistant API
api:
  encryption:
    key: "blah"

ota:
  password: "blah"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Upstairs-Grid-Eye"
    password: "blah"

captive_portal:

i2c:
  sda:  26
  scl:  32
  scan: True
  frequency: 400kHz


text_sensor:
  - platform: custom
    lambda: |-
      auto amg8833 = new AMG8833CameraComponent();
      App.register_component(amg8833);
      return {amg8833};
    text_sensors:
      - name: "Sensor Pixels"

binary_sensor:
  - platform: custom
    lambda: |-
      auto amg8833 = new AMG8833PersonDetectedComponent();
      App.register_component(amg8833);
      return {amg8833};
    binary_sensors:
      name: "Person Detected"
      device_class: occupancy

sensor:
  - platform: custom
    lambda: |-
      auto amg8833 = new AMG8833Component();
      App.register_component(amg8833);
      return {amg8833->sensor_temperature, amg8833->max_temperature, amg8833->min_temperature, amg8833->avg_temperature, amg8833->min_index, amg8833->max_index};
    sensors:
      - name: "Thermal Sensor Temperature"
        unit_of_measurement: °C
        device_class: temperature
        accuracy_decimals: 2

      - name: "Thermal Sensor Max"
        unit_of_measurement: °C
        device_class: temperature
        accuracy_decimals: 2

      - name: "Thermal Sensor Min"
        unit_of_measurement: °C
        device_class: temperature
        accuracy_decimals: 2

      - name: "Thermal Sensor Avg"
        unit_of_measurement: °C
        device_class: temperature
        accuracy_decimals: 2

      - name: "Thermal Sensor Min Index"
        accuracy_decimals: 0

      - name: "Thermal Sensor Max Index"
        accuracy_decimals: 0
```

**Comment by TheRealWaldo** [comment]
I believe you'd be looking for ESPHome sensor filters: https://esphome.io/components/binary_sensor/index.html

**Comment by Zipties** [comment]
> I believe you'd be looking for ESPHome sensor filters: https://esphome.io/components/binary_sensor/index.html

Yep, that did the trick, appreciate it! That stabilized everything a lot, but I am still trying to figure out my tuning options. I found where to change the temperature differential, is there anything else that may be useful? Would it be better to get readings from a climate sensor in the room? 


Delay code for googlers
```
binary_sensor:
  - platform: custom
    lambda: |-
      auto amg8833 = new AMG8833PersonDetectedComponent();
      App.register_component(amg8833);
      return {amg8833};
    binary_sensors:
      name: "Person Detected"
      device_class: occupancy
      filters:
        -  delayed_off: 1.5s
```

**Comment by TheRealWaldo** [comment]
The challenge there is that ambient air temperature and surface temperature are two very different things.

The best thing you can do is carefully aim the sensor to avoid typically hot and cold spots and work off of purely differential, unless you have another IR sensor for a reference surface temperature.

**Comment by Zipties** [comment]
Yeah, aiming is proving to be a pita with such a narrow FOV, even my monitor sets it off. It seems to be pretty time of the day sensitive based on how much sun the room is getting, even with the blinds closed. Morning 3 degree differential works well, but then sticks as triggered until I change it back up to 4 in the afternoon

What do the min/max index values represent?

![CleanShot 2022-06-01 at 18 33 45](https://user-images.githubusercontent.com/11506166/171518435-073fc460-9ae1-4789-b508-d2c48dac23d6.jpg)

**Comment by TheRealWaldo** [comment]
If you think of the output from the sensor as serial, it's the position of the high/low values in that output.
